stan/math/prim/arr/fun/vec_concat.hpp

```diff
@@ -0,0 +1,48 @@
+#ifndef STAN_MATH_PRIM_ARR_FUN_VEC_CONCAT_HPP
+#define STAN_MATH_PRIM_ARR_FUN_VEC_CONCAT_HPP
+
+#include <type_traits>
+#include <iterator>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+
+template <typename T>
+void concat_helper(std::vector<T>& l, const std::vector<T>& r) {
+  l.insert(l.end(), r.begin(), r.end());
+}
+
+template <typename T>
+void concat_helper(std::vector<T>& l, std::vector<T>&& r) {
+  l.insert(l.end(), std::make_move_iterator(r.begin()),
+           std::make_move_iterator(r.end()));
+}
+}  // namespace internal
+
+/**
+ * Gets the event stack from a vector of events and other arguments.
+ * @param v1 A event stack to roll up.
+ * @param args variadic arcs passed down to the next recursion.
+ * @tparam Args Types for variadic.
+ * @return Vector of OpenCL events
+ */
+template <typename T, typename... Args>
+std::vector<T> vec_concat(const std::vector<T> v1, Args&&... args) {
+  std::size_t s = v1.size();
+  // Hacks to expand the variadic template
+  auto dummy1 = {s += args.size()...};
```

Comment:
This is pretty confusing. I found the stackoverflow question this came from and I still don't really get why this works - seems like it's taking advantage of some weird spec interactions. I am hopeful we won't ever have giant vectors such that we don't need to worry about the .reserve() call, in which case could we do something like vec_concat here: https://ideone.com/5Bbo7r ?

Or didn't we have something simpler before? I don't remember.

Reply:
Yeah tbh I was trying to learn more about perfect forwarding and based this off some code I found on stackoverflow. If the code here has an odd smell I'm fine with moving back to the other code. Don't think it will be that different in terms of speed